sys/net/application_layer/asymcute/asymcute.c

```diff
@@ -0,0 +1,965 @@
+            _on_puback(con, con->rxbuf, len);
+            break;
+        case MQTTSN_SUBACK:
+            _on_suback(con, con->rxbuf, len);
+            break;
+        case MQTTSN_UNSUBACK:
+            _on_unsuback(con, con->rxbuf, len);
+            break;
+        default:
+            break;
+    }
+}
+
+void *_listener(void *arg)
+{
+    asymcute_con_t *con = (asymcute_con_t *)arg;
+
+    /* create a socket for this listener, using an ephemeral port */
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+    if (sock_udp_create(&con->sock, &local, NULL, 0) != 0) {
+        LOG_ERROR("[asymcute] error creating listener socket\n");
+        return NULL;
+    }
+
+    while (1) {
+        sock_udp_ep_t remote;
+        int n = sock_udp_recv(&con->sock, con->rxbuf, ASYMCUTE_BUFSIZE,
+                              SOCK_NO_TIMEOUT, &remote);
+        if (n > 0) {
+            _on_data(con, (size_t)n, &remote);
+        }
+    }
+
+    /* should never be reached */
+    return NULL;
+}
+
+void *_handler(void *arg)
+{
+    (void)arg;
+    event_queue_init(&_queue);
+    event_loop(&_queue);
+    /* should never be reached */
+    return NULL;
+}
+
+int asymcute_listener_run(asymcute_con_t *con, char *stack, size_t stacksize,
+                          char priority, asymcute_evt_cb_t callback)
+{
+    /* make sure con is not running */
+    assert(con);
+    assert((priority > 0) && (priority < THREAD_PRIORITY_IDLE - 1));
+    assert(callback);
+
+    int ret = ASYMCUTE_OK;
+
+    /* make sure the connection context is not already used */
+    mutex_lock(&con->lock);
+    if (con->state != UNINITIALIZED) {
+        ret = ASYMCUTE_BUSY;
+        goto end;
+    }
+
+    /* initialize the connection context */
+    memset(con, 0, sizeof(asymcute_con_t));
+    random_bytes((uint8_t *)&con->last_id, 2);
+    event_callback_init(&con->keepalive_evt, _on_keepalive_evt, con);
+    event_timeout_init(&con->keepalive_timer, &_queue, &con->keepalive_evt.super);
+    con->keepalive_retry_cnt = ASYMCUTE_N_RETRY;
+    con->state = NOTCON;
+    con->user_cb = callback;
+
+    /* start listener thread */
+    thread_create(stack,
+                  stacksize,
+                  priority,
+                  THREAD_CREATE_WOUT_YIELD,
+                  _listener,
+                  con,
+                  "asymcute_listener");
+
+end:
+    mutex_unlock(&con->lock);
+    return ret;
+}
+
+void asymcute_handler_run(void)
+{
+    thread_create(_stack, sizeof(_stack), ASYMCUTE_HANDLER_PRIO,
+                  0, _handler, NULL, "asymcute_main");
+}
+
+int asymcute_topic_init(asymcute_topic_t *topic, const char *topic_name,
+                        uint16_t topic_id)
+{
+    assert(topic);
+
+    size_t len = strlen(topic_name);
+
+    if (asymcute_topic_is_reg(topic)) {
+        return ASYMCUTE_REGERR;
+    }
+    if ((topic_name && ((len == 0) || (len > ASYMCUTE_TOPIC_MAXLEN))) ||
+        (!topic_name && ((topic_id == 0) || (topic_id == UINT16_MAX)))) {
```

Comment:
`cppcheck` reports the check for `!topic_name` to be redundant, and since you can, due to lazy evaluation only land here when `topic_name` is true in the line above, I say its right. So please remove.

Reply:
Ah, it's your usage of `strlen(topic_name)` in https://github.com/RIOT-OS/RIOT/pull/9464/files#diff-640646a2b9d2e083e6a60251fdfd4d99R652.